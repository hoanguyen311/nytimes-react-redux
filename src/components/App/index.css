body {
  margin: 0;
  padding: 0;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, "Courier New",
    monospace;
}

.app-content {
  padding: 0 50px;
}

.app-logo {
  height: 50px;
  float: left;
}


.app-header {
  background: #FFF;
  border-bottom: 1px solid #999;
  justify-content: center;
  display: flex;
}


.app-headline {
  /* margin-bottom: 0; */
  display: inline-block;
}
