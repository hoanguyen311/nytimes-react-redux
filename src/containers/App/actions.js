export const FETCH_NEWS_DATA = 'FETCH_NEWS_DATA';
export const FETCH_NEWS_DATA_SUCCESS = 'FETCH_NEWS_DATA_SUCCESS';
export const FETCH_NEWS_DATA_FAILURE = 'FETCH_NEWS_DATA_FAILURE';


export function fetchNewsData() {
  return {
    type: FETCH_NEWS_DATA,
  };
}

export function fetchNewsDataSuccess(payload) {
  return {
    type: FETCH_NEWS_DATA_SUCCESS,
    payload
  };
}

export function fetchNewsDataFailure(error) {
  return {
    type: FETCH_NEWS_DATA_FAILURE,
    error
  };
}
