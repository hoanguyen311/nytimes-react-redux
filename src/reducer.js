import { combineReducers } from 'redux';
import app from '@containers/App/reducer';
import newsList from '@containers/NewsList/reducer';

export default combineReducers({
  app,
  newsList
});
