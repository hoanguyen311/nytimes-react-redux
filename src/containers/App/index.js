import { connect } from 'react-redux';
import App from '@components/App';
import { fetchNewsData as fetchData } from './actions';

const mapDispatchToProps = {
  fetchData
};

const mapStateToProps = (state) => ({
  loading: state.app.loading
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
