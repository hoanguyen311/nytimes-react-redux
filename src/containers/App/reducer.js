import { FETCH_NEWS_DATA, FETCH_NEWS_DATA_SUCCESS, FETCH_NEWS_DATA_FAILURE } from './actions';

export default function(state = initialState, { type, payload, error }) {
  switch(type) {
    case FETCH_NEWS_DATA:
      return {
        ...state,
        loading: true
      };

    case FETCH_NEWS_DATA_SUCCESS:
      return {
        ...state,
        loading: false
      };

    case FETCH_NEWS_DATA_FAILURE:
      return {
        ...state,
        error,
        loading: false
      };

    default:
      return state;
  }
}

export const initialState = {
  loading: true,
  error: null
};
