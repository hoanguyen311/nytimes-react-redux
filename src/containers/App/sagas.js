
import { takeLatest, call, put } from 'redux-saga/effects';
import request from '@/utils';
import { FETCH_NEWS_DATA, fetchNewsDataSuccess, fetchNewsDataFailure } from './actions';

export const API_URL = 'https://api.nytimes.com/svc/search/v2/articlesearch.json?api-key=5763846de30d489aa867f0711e2b031c&q=singapore&page=0';
export function* fetchNewsData() {
  try {
    const { status, response } = yield call(request, API_URL);

    if (status === 'OK') {
      yield put(fetchNewsDataSuccess(response));
      return;
    }

    yield put(fetchNewsDataFailure(response));
  } catch(error) {
    yield put(fetchNewsDataFailure(error));
  }
}

export default function* () {
  yield takeLatest(FETCH_NEWS_DATA, fetchNewsData);
}
