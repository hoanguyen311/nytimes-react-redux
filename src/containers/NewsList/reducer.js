import { combineReducers } from 'redux';
import { format } from 'date-fns';
import { FETCH_NEWS_DATA_SUCCESS } from '@containers/App/actions';

export const initialState = {
  allIds: [],
  byId: {}
};

export function normalizeItem(item) {
  let image = null;

  if (item.multimedia.length && item.multimedia[0].url) {
    image = `https://www.nytimes.com/${item.multimedia[0].url}`;
  }

  return {
    ...item,
    id: item._id,
    headline: item.headline.main,
    image,
    formattedDate: item.pub_date ? format(item.pub_date, 'MM/DD/YYYY') : null
  };
}

export function byId(state = initialState.byId, { type, payload }) {
  switch(type) {
    case FETCH_NEWS_DATA_SUCCESS:
      return payload.docs.reduce((results, doc) => {
        return {
          ...results,
          [doc._id]: normalizeItem(doc)
        };
      }, {});

    default:
      return state;
  }
}

export function allIds(state = initialState.allIds, { type, payload }) {
  switch(type) {
    case FETCH_NEWS_DATA_SUCCESS:
      return state.concat(payload.docs.map(doc => doc._id));

    default:
      return state;
  }
}

export function getAll({ allIds, byId }) {
  return allIds.map((id) => byId[id]);
}

export default combineReducers({
  byId,
  allIds
});
