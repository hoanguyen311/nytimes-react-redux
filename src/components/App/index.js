import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Layout } from 'antd';
import { BrowserRouter as Router, Route, Link } from 'react-router-dom';
import NewsList from '@containers/NewsList';
import Article from '@containers/Article';
import logo from './nytlogo.png';
import './index.css';

const { Header, Content } = Layout;

class App extends Component {
  static propTypes = {
    fetchData: PropTypes.func,
  };

  static defaultProps = {
    fetchData: null,
  };

  componentDidMount() {
    const { fetchData } = this.props;
    fetchData && fetchData();
  }

  render() {
    return (
      <Router>
        <Layout>
          <Header className="app-header">
            <h1 className="app-headline">
              <Link to="/">
                <img src={logo} alt="The New York Times" className="app-logo" />
              </Link>
            </h1>
          </Header>
            <Content className="app-content">
              <Route exact path="/" component={NewsList} />
              <Route exact path="/article/:id" render={({match}) => <Article id={match.params.id} />} />
            </Content>
        </Layout>
      </Router>
    );
  }
}

export default App;
