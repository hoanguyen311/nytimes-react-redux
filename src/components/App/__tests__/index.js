import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import reducer from '@/reducer';
import { fetchNewsData, fetchNewsDataSuccess, fetchNewsDataFailure } from '@/containers/App/actions';
import App from '../index';

let createStore = null;

beforeEach(() => {
  createStore = configureStore();
});

it('fetchNewsData', () => {
  const store = createStore(reducer(undefined, fetchNewsData()));

  const component = renderer.create(
    <Provider store={store}>
      <App />
    </Provider>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('fetchNewsDataSuccess', () => {
  const action = fetchNewsDataSuccess({
    docs: []
  });
  const store = createStore(reducer(undefined, action));

  const component = renderer.create(
    <Provider store={store}>
      <App />
    </Provider>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('fetchNewsDataFailure', () => {
  const action = fetchNewsDataFailure({
    message: ''
  });
  const store = createStore(reducer(undefined, action));

  const component = renderer.create(
    <Provider store={store}>
      <App />
    </Provider>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

