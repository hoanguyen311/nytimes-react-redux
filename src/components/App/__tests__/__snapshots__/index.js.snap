// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchNewsData 1`] = `
<div
  className="ant-layout"
>
  <div
    className="app-header ant-layout-header"
  >
    <h1
      className="app-headline"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <img
          alt="The New York Times"
          className="app-logo"
          src="nytlogo.png"
        />
      </a>
    </h1>
  </div>
  <div
    className="app-content ant-layout-content"
  >
    <div
      className="ant-list ant-list-vertical ant-list-lg ant-list-split"
    >
      <div
        className="ant-spin-nested-loading"
        style={null}
      >
        <div
          className="ant-spin-container"
        >
          <div
            className="ant-list-item"
          >
            <div
              className="ant-list-item-content ant-list-item-content-single"
            >
              <div
                className="ant-skeleton ant-skeleton-active"
              >
                <div
                  className="ant-skeleton-content"
                >
                  <h3
                    className="ant-skeleton-title"
                    style={
                      Object {
                        "width": "38%",
                      }
                    }
                  />
                  <ul
                    className="ant-skeleton-paragraph"
                  >
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": "61%",
                        }
                      }
                    />
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div
            className="ant-list-item"
          >
            <div
              className="ant-list-item-content ant-list-item-content-single"
            >
              <div
                className="ant-skeleton ant-skeleton-active"
              >
                <div
                  className="ant-skeleton-content"
                >
                  <h3
                    className="ant-skeleton-title"
                    style={
                      Object {
                        "width": "38%",
                      }
                    }
                  />
                  <ul
                    className="ant-skeleton-paragraph"
                  >
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": "61%",
                        }
                      }
                    />
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div
            className="ant-list-item"
          >
            <div
              className="ant-list-item-content ant-list-item-content-single"
            >
              <div
                className="ant-skeleton ant-skeleton-active"
              >
                <div
                  className="ant-skeleton-content"
                >
                  <h3
                    className="ant-skeleton-title"
                    style={
                      Object {
                        "width": "38%",
                      }
                    }
                  />
                  <ul
                    className="ant-skeleton-paragraph"
                  >
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": "61%",
                        }
                      }
                    />
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div
            className="ant-list-item"
          >
            <div
              className="ant-list-item-content ant-list-item-content-single"
            >
              <div
                className="ant-skeleton ant-skeleton-active"
              >
                <div
                  className="ant-skeleton-content"
                >
                  <h3
                    className="ant-skeleton-title"
                    style={
                      Object {
                        "width": "38%",
                      }
                    }
                  />
                  <ul
                    className="ant-skeleton-paragraph"
                  >
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": "61%",
                        }
                      }
                    />
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div
            className="ant-list-item"
          >
            <div
              className="ant-list-item-content ant-list-item-content-single"
            >
              <div
                className="ant-skeleton ant-skeleton-active"
              >
                <div
                  className="ant-skeleton-content"
                >
                  <h3
                    className="ant-skeleton-title"
                    style={
                      Object {
                        "width": "38%",
                      }
                    }
                  />
                  <ul
                    className="ant-skeleton-paragraph"
                  >
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": "61%",
                        }
                      }
                    />
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div
            className="ant-list-item"
          >
            <div
              className="ant-list-item-content ant-list-item-content-single"
            >
              <div
                className="ant-skeleton ant-skeleton-active"
              >
                <div
                  className="ant-skeleton-content"
                >
                  <h3
                    className="ant-skeleton-title"
                    style={
                      Object {
                        "width": "38%",
                      }
                    }
                  />
                  <ul
                    className="ant-skeleton-paragraph"
                  >
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": "61%",
                        }
                      }
                    />
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div
            className="ant-list-item"
          >
            <div
              className="ant-list-item-content ant-list-item-content-single"
            >
              <div
                className="ant-skeleton ant-skeleton-active"
              >
                <div
                  className="ant-skeleton-content"
                >
                  <h3
                    className="ant-skeleton-title"
                    style={
                      Object {
                        "width": "38%",
                      }
                    }
                  />
                  <ul
                    className="ant-skeleton-paragraph"
                  >
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": "61%",
                        }
                      }
                    />
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div
            className="ant-list-item"
          >
            <div
              className="ant-list-item-content ant-list-item-content-single"
            >
              <div
                className="ant-skeleton ant-skeleton-active"
              >
                <div
                  className="ant-skeleton-content"
                >
                  <h3
                    className="ant-skeleton-title"
                    style={
                      Object {
                        "width": "38%",
                      }
                    }
                  />
                  <ul
                    className="ant-skeleton-paragraph"
                  >
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": "61%",
                        }
                      }
                    />
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div
            className="ant-list-item"
          >
            <div
              className="ant-list-item-content ant-list-item-content-single"
            >
              <div
                className="ant-skeleton ant-skeleton-active"
              >
                <div
                  className="ant-skeleton-content"
                >
                  <h3
                    className="ant-skeleton-title"
                    style={
                      Object {
                        "width": "38%",
                      }
                    }
                  />
                  <ul
                    className="ant-skeleton-paragraph"
                  >
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": "61%",
                        }
                      }
                    />
                  </ul>
                </div>
              </div>
            </div>
          </div>
          <div
            className="ant-list-item"
          >
            <div
              className="ant-list-item-content ant-list-item-content-single"
            >
              <div
                className="ant-skeleton ant-skeleton-active"
              >
                <div
                  className="ant-skeleton-content"
                >
                  <h3
                    className="ant-skeleton-title"
                    style={
                      Object {
                        "width": "38%",
                      }
                    }
                  />
                  <ul
                    className="ant-skeleton-paragraph"
                  >
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": undefined,
                        }
                      }
                    />
                    <li
                      style={
                        Object {
                          "width": "61%",
                        }
                      }
                    />
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`fetchNewsDataFailure 1`] = `
<div
  className="ant-layout"
>
  <div
    className="app-header ant-layout-header"
  >
    <h1
      className="app-headline"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <img
          alt="The New York Times"
          className="app-logo"
          src="nytlogo.png"
        />
      </a>
    </h1>
  </div>
  <div
    className="app-content ant-layout-content"
  >
    <div
      className="ant-list ant-list-vertical ant-list-lg ant-list-split"
    >
      <div
        className="ant-spin-nested-loading"
        style={null}
      >
        <div
          className="ant-spin-container"
        >
          <div
            className="ant-list-empty-text"
          >
            No data
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`fetchNewsDataSuccess 1`] = `
<div
  className="ant-layout"
>
  <div
    className="app-header ant-layout-header"
  >
    <h1
      className="app-headline"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <img
          alt="The New York Times"
          className="app-logo"
          src="nytlogo.png"
        />
      </a>
    </h1>
  </div>
  <div
    className="app-content ant-layout-content"
  >
    <div
      className="ant-list ant-list-vertical ant-list-lg ant-list-split"
    >
      <div
        className="ant-spin-nested-loading"
        style={null}
      >
        <div
          className="ant-spin-container"
        >
          <div
            className="ant-list-empty-text"
          >
            No data
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
